File: client/tftp_client.cpp
#include "../server/main.h"

#include "tftp_packet.h"
#include "tftp_client.h"

using namespace std;

TFTPClient::TFTPClient(char* ip, int port) {

	TFTP_Packet packet;

	server_ip = ip;
	server_port = port;

	//- standartines reiksmes

	socket_descriptor = -1;

	//- wsa

	#ifdef WIN32

		/* edited */

		WSADATA wsaData;

		WORD wVersionRequested = MAKEWORD(2, 2);

		int err = WSAStartup(wVersionRequested, &wsaData);

		if (err != 0) {

			throw new ETFTPSocketInitialize;
		    
		}

	#endif

}

int TFTPClient::connectToServer() {

	cout << "Connecting to " << server_ip << " on port " << server_port << endl;

    socket_descriptor = socket(PF_INET, SOCK_STREAM, 0);

    if (socket_descriptor == -1) {

        throw new ETFTPSocketCreate;

    }

    DEBUGMSG("Socket created");

    client_address.sin_family = AF_INET;
	client_address.sin_port = htons(server_port);	//- taip pat turi buti ir serveryje!
    client_address.sin_addr.s_addr = inet_addr(this->server_ip);

    #ifdef WIN32
        //memset(client_address.sin_zero, 0, sizeof(client_address.sinzero);
        //- suvienodinam SOCKADDR_IN ir SOCKADDR strukturu dydzius
    #endif

    connection = connect(socket_descriptor, (const struct sockaddr *)&client_address, sizeof(client_address));

    if (connection != 0) {

        //cout << "Unable to connect to an address\n";
        return -1;
        
    }

    DEBUGMSG("Successfully connected");

    return 1;

}

int TFTPClient::sendBuffer(char *buffer) {

    return send(socket_descriptor, buffer, (int)strlen(buffer), 0);

}

int TFTPClient::sendPacket(TFTP_Packet* packet) {

	return send(socket_descriptor, (char*)packet->getData(), packet->getSize(), 0);

}

bool TFTPClient::getFile(char* filename,char* destination) {

	TFTP_Packet packet_rrq, packet_ack;
	ofstream file(destination, ifstream::binary);

	char buffer[TFTP_PACKET_DATA_SIZE];

	packet_rrq.createRRQ(filename);

	sendPacket(&packet_rrq);

	int last_packet_no = 1;
	int wait_status;
	int timeout_count = 0;

	while (true) {

		wait_status = waitForPacketData(last_packet_no, TFTP_CLIENT_SERVER_TIMEOUT);
		
		if (wait_status == TFTP_CLIENT_ERROR_PACKET_UNEXPECTED) {
			
			received_packet.dumpData();
			file.close();

			return false;

		} else if (wait_status == TFTP_CLIENT_ERROR_TIMEOUT) {

			//- nebesulaukem paketo

			timeout_count++;

			if (timeout_count < 2) { //- kadangi tai pirmas timeout`as, tai bandom pasiusti paskutini ACK

				sendPacket(&packet_ack);
				continue;

			} else {

				cout << "Connection timeout" << endl;

				file.close();

				return false;

			}

		}

		if (last_packet_no != received_packet.getNumber()) {
			//- paketas kazkur paklydo!

			/* TFTP recognizes only one error condition that does not cause
			   termination, the source port of a received packet being incorrect.
			   In this case, an error packet is sent to the originating host. */

			/* Taip negali nutikti, nes pas mus naudojamas ACK Lock`as */

			cout << "This should not happen!" << endl; //- aisku kada kadanors tai vis tiek atsitiks :)

		} else {

			//- paketas tvarkoj
			received_packet.dumpData();

			last_packet_no++;

			//- jei tai susitvarkes timeoutinis paketas, tai pasidziaukim ir leiskim tai pakartot
			if (timeout_count == 1) {
				timeout_count = 0;
			}

			if (received_packet.copyData(4, buffer, TFTP_PACKET_DATA_SIZE)) {

				file.write(buffer, received_packet.getSize() - 4);

				//- tirkinam, ar gauti duomenis yra mazesni nei buferio dydis
				//- jei taip, tai sis paketas buvo paskutinis
				//- A data packet of less than 512 bytes signals termination of a transfer.

				if (received_packet.getSize() - 4 < TFTP_PACKET_DATA_SIZE) {

					/* The host acknowledging the final DATA packet may terminate its side
					   of the connection on sending the final ACK. */

					packet_ack.createACK((last_packet_no - 1));

					if (sendPacket(&packet_ack)) {

						break;

					}

				} else {

					//- ne paskutinis, tai siunciam ACK
					//- Each data packet contains one block of data, and must be acknowledged by 
					//- an acknowledgment packet before the next packet can be sent.
					
					packet_ack.createACK((last_packet_no - 1)); //- siunciam toki paketo numeri, kuri gavom paskutini
					
					sendPacket(&packet_ack);

					DEBUGMSG("Acknoledgement sent");

				}

			}

		}

	}

	file.close();

	return true;

}

int TFTPClient::waitForPacket(TFTP_Packet* packet, int timeout_ms) {

	packet->clear();

	FD_SET fd_reader;		  // soketu masyvo struktura
	timeval connection_timer; // laiko struktura perduodama select()

	connection_timer.tv_sec = timeout_ms / 1000; // s
	connection_timer.tv_usec = 0; // neveikia o.0 timeout_ms; // ms 

	FD_ZERO(&fd_reader);

	// laukiam, kol bus ka nuskaityti
	FD_SET(socket_descriptor, &fd_reader);

	int select_ready = select(0, &fd_reader, NULL, &fd_reader, &connection_timer);

	if (select_ready == -1) {

		cout << "Error in select(), no: " << WSAGetLastError() << endl;
		return TFTP_CLIENT_ERROR_SELECT;

	} else if (select_ready == 0) {

		DEBUGMSG("Timeout");
		return TFTP_CLIENT_ERROR_TIMEOUT;

	}

	//- turim sekminga event`a

	int receive_status;

	receive_status = recv(socket_descriptor, (char*)packet->getData(), TFTP_PACKET_MAX_SIZE, 0);

	if (receive_status == 0) {
		cout << "Connection was closed by server\n";
		return TFTP_CLIENT_ERROR_CONNECTION_CLOSED;
    }

	if (receive_status == SOCKET_ERROR)	{
		DEBUGMSG("recv() error in waitForPackage()");
		return TFTP_CLIENT_ERROR_RECEIVE;
	}

	//- receive_status - gautu duomenu dydis
	
	packet->setSize(receive_status);

	return TFTP_CLIENT_ERROR_NO_ERROR;

}

bool TFTPClient::waitForPacketACK(int packet_number, int timeout_ms) {

	TFTP_Packet received_packet;

	if (waitForPacket(&received_packet, 1000)) {

		if (received_packet.isError()) {

			errorReceived(&received_packet);

			return false;

		}

		if (received_packet.isACK()) {

			return true;

		}

		if (received_packet.isData()) {

			return true;

		}

	}

	return true;

}

int TFTPClient::waitForPacketData(int packet_number, int timeout_ms) {

	int wait_status = waitForPacket(&received_packet, timeout_ms);

	if (wait_status == TFTP_CLIENT_ERROR_NO_ERROR) {

		if (received_packet.isError()) {

			errorReceived(&received_packet);

			return TFTP_CLIENT_ERROR_PACKET_UNEXPECTED;

		}

		if (received_packet.isData()) {
			
			return TFTP_CLIENT_ERROR_NO_ERROR;

		}

	}

	return wait_status;

}

bool TFTPClient::sendFile(char* filename, char* destination) {
	
	TFTP_Packet packet_wrq, packet_data;
	ifstream file(filename, ifstream::binary);
	char memblock[TFTP_PACKET_DATA_SIZE];

	if (!file.is_open() || !file.good()) {

		cout << "Unable to open file " << filename << endl;

		return false;

	}

	char buffer[TFTP_PACKET_DATA_SIZE];

	packet_wrq.createWRQ(destination);
	packet_wrq.dumpData();

	sendPacket(&packet_wrq);

	int last_packet_no = 0;
	int last_error = 0;

	while (true) {

		if (waitForPacketACK(++last_packet_no, TFTP_CLIENT_SERVER_TIMEOUT)) {

			file.read(memblock, TFTP_PACKET_DATA_SIZE);

			packet_data.createData(last_packet_no, (char*)memblock, file.gcount());

			sendPacket(&packet_data);

			if (file.eof()) {

				break;

			}

		} else {

			cout << "Server has timed out" << endl;
			file.close();
			return false;

		}
		
	}

	file.close();

	return true;

}

void TFTPClient::errorReceived(TFTP_Packet* packet) {

	int error_code = packet->getWord(2);

	cout << "Error! Error code: " << error_code << endl;
	cout << "Error message: ";
	
	switch (error_code) {

		case 1: cout << TFTP_ERROR_1; break;
		case 2: cout << TFTP_ERROR_2; break;
		case 3: cout << TFTP_ERROR_3; break;
		case 4: cout << TFTP_ERROR_4; break;
		case 5: cout << TFTP_ERROR_5; break;
		case 6: cout << TFTP_ERROR_6; break;
		case 7: cout << TFTP_ERROR_7; break;
		case 0: 
		default: cout << TFTP_ERROR_0; break;

	}

	cout << endl;

	TFTPClient::~TFTPClient();

}

TFTPClient::~TFTPClient() {

    if (socket_descriptor != -1) {

        #ifdef WIN32

            closesocket(socket_descriptor);
            WSACleanup();
            
        #else

            close(socket_descriptor);

        #endif

    }

}

void DEBUGMSG(char *msg) {
    #ifdef DEBUG
	std::cout << msg << "\n";
    #endif
}

File: java/src/tftpclient/TFTPClient.java
package tftpclient;

import java.net.*;
import java.io.*;

import tftp.*;
import tftpserver.TFTPServerClient;
import tftpserver.TFTPServerClient.TClientRequest;


public class TFTPClient extends Thread {
	
	public int server_port;
	public String server_ip;
	public Socket server_socket;
	
	public BufferedOutputStream out;
	public BufferedInputStream in;
	public DataInputStream fin;	
	public DataOutputStream fout;	
	
	public TFTPServerClient.TClientRequest request;
	
	public TFTPClient(int port, String ip) {
	
		server_port = port;
		server_ip = ip;
		
		try {
			
			server_socket = new Socket(server_ip, server_port);
			
		} catch (UnknownHostException unknownHost) {
			
			TFTPUtils.fatalError("Unknown host " + server_ip);
			
		} catch (IOException ioException) {
			
			TFTPUtils.fatalError("Unable to create a socket");
                
        }
        
	}
	
	public boolean sendFile(String filename, String dest) {
		
		TFTPPacket packet_wrq, packet_ack, packet_data;
		int bytes_read;
		int packet_no;
		byte[] data = new byte[TFTPPacket.TFTP_PACKET_DATA_SIZE];
		
		try {
			
			fin = new DataInputStream(new FileInputStream("D:\\Users\\Vilius\\Desktop\\tftp\\java\\src\\tftpclient\\" + filename));
			
		} catch (Exception e) {

			TFTPUtils.fatalError("Unable to open source file: " + filename);

			return false;

		}
		
		request = TClientRequest.WRQ;
		
		if (server_socket.isConnected()) {
			TFTPUtils.puts("Connected to " + server_ip + " on port " + server_port);			
		} else {
			TFTPUtils.fatalError("Disconnected unexpectedly");
		}
		
        try {
        	
	        out = new BufferedOutputStream(server_socket.getOutputStream());
	        in = new BufferedInputStream(server_socket.getInputStream());
	        
	        packet_wrq = new TFTPPacket();
	        	
	        packet_wrq.createWRQ(dest);
	        packet_wrq.dumpData();
	        	
	        TFTPUtils.puts("Sending write request to server");
	        packet_wrq.sendPacket(out);
	        
	        // wait for ack
	        packet_ack = new TFTPPacket();
	        packet_data = new TFTPPacket();
	        
	        packet_no = 0;
	        
	        mainloop:
	        while (true) {
	        	
	        	TFTPUtils.puts("Waiting for ACK");
		   		
		   		if (!packet_ack.getPacket(in)) {
	        	
		   			TFTPUtils.puts("End of file reached");
		   			break;
		   			
		   		}
		   		
		   		if (packet_ack.isACK()) {
		   			
		   			if (packet_ack.getPacketNumber() != (char)packet_no) {
		   				TFTPUtils.puts("WRONG ORDER");
		   			}
		   			
		   			TFTPUtils.puts("ACK received, sending next DATA packet");
		   			
		   			for (int i = 0; i < TFTPPacket.TFTP_PACKET_DATA_SIZE; i++) data[i] = 0;
		   			
		   			bytes_read = 0;
		   	   		
		   			try {
		   				
		   				bytes_read = fin.read(data, 0, TFTPPacket.TFTP_PACKET_DATA_SIZE);
		   				
		   				if (bytes_read == -1) {
		   					
		   					TFTPUtils.puts("EOF in sendFile()");
		   					
		   	   	   		} else {
		   	   	   			
		   	   	   			packet_no++;
		   	   	   			
		   	   	   			packet_data.createData(packet_no, data, bytes_read);
		   	   	   			packet_data.sendPacket(out);
		   	   	   			
		   	   	   			TFTPUtils.puts("Data packet sent");
		   	   	   			packet_data.dumpData();
		   	   	   			
		   	   	   		}
		   				
		   				if (bytes_read < TFTPPacket.TFTP_PACKET_DATA_SIZE) {
		   					
		   					//- this is our last packet
		   					
		   					TFTPUtils.puts("File was successfully sent!");
		   					disconnect();
		   					
		   					break mainloop;
		   					
		   				}
		   				
		   	   		} catch (Exception e) {
		   	   			TFTPUtils.puts("Exception in sendFile() isACK()");		   	   			
		   	   		}
		   			
		   		} else if (packet_ack.isError()) {
		   			
		   			try {
		   				TFTPUtils.puts("Error packet received with message: " + packet_ack.getString(4, packet_ack.getSize()));
		   			} catch (Exception e) {
					}
		   			disconnect();
		   			
		   			break;
		   			
		   		} else {
		   			
		   			TFTPUtils.puts("Unexpected packet");
		   			disconnect();
		   			
		   			break;
		   			
		   		}
		   		
	        }
	        	
        } catch (IOException ioException) {
        	
            TFTPUtils.puts("IO Exception in thread: " + ioException.getMessage());
        	
        }
		
		return true;
		
	}
	
	public boolean getFile(String filename, String dest) {
		
		TFTPPacket packet_rrq, packet_ack, packet_data;
		int packet_no;
		
		File file_exists = null;

		try {

			file_exists = new File(dest);

		} catch (Exception e) {

			TFTPUtils.fatalError(e.getMessage());

		}

		if (file_exists != null && file_exists.exists()) {

			TFTPUtils.fatalError("Unable to download file to " + dest + ". Destination already exists in " + file_exists.getAbsolutePath());
			
		}
		
		try {
			
			fout = new DataOutputStream(new FileOutputStream(dest));
			
		} catch (Exception e) {

			TFTPUtils.fatalError("Unable to download file to specified destination: " + dest);

			return false;

		}
		
		request = TClientRequest.RRQ;
		
		if (server_socket.isConnected()) {
			TFTPUtils.puts("Connected to " + server_ip + " on port " + server_port);			
		} else {
			TFTPUtils.fatalError("Disconnected unexpectedly");
		}
		
        try {
        	
	        out = new BufferedOutputStream(server_socket.getOutputStream());
	        in = new BufferedInputStream(server_socket.getInputStream());
	        
	        packet_rrq = new TFTPPacket();
	        	
	        packet_rrq.createRRQ(filename);
	        packet_rrq.dumpData();
	        	
	        TFTPUtils.puts("Sending read request to server");
	        packet_rrq.sendPacket(out);
	        
	        // wait for data packet
	        packet_ack = new TFTPPacket();
	        packet_data = new TFTPPacket();
	        
	        packet_no = 0;
	        
	        mainloop:
	        while (true) {
	        	
	        	TFTPUtils.puts("Waiting for DATA packet");
		   		
		   		if (!packet_data.getPacket(in)) {
	        	
		   			TFTPUtils.puts("End of file reached");
		   			break;
		   			
		   		}
		   		
		   		//- gavom duomenu paketa
		   		if (packet_data.isData()) {
		   			
		   			if (packet_data.getPacketNumber() != (char)packet_no) {
		   				TFTPUtils.puts("WRONG ORDER");
		   			}
		   			
					fout.write(packet_data.getData(4));

					TFTPUtils.puts("DATA received, sending ACK packet");
					packet_ack.createACK(packet_data.getPacketNumber());
					packet_ack.sendPacket(out);
						
					if (packet_data.getSize() < 4 + TFTPPacket.TFTP_PACKET_DATA_SIZE) {
						
						TFTPUtils.puts("File transferred");
						break;
						
					}
					
		   		} else if (packet_data.isError()) {
						
		   			try {
		   				TFTPUtils.puts("Error packet received with message: " + packet_data.getString(4, packet_data.getSize()));
		   			} catch (Exception e) {
					}
		   			disconnect();
		   			
		   		} else {
		   			
		   			TFTPUtils.puts("Unexpected packet");
		   			disconnect();
		   			
		   			break mainloop;
		   			
		   		}
		   		
	        }
	        	
        } catch (IOException ioException) {
        	
            TFTPUtils.puts("IO Exception in thread: " + ioException.getMessage());
        	
        }
		
		return true;
		
	}
	
	public void disconnect() {
		
		TFTPUtils.puts("Disconnected from server");
		
		try {
			server_socket.close();
			out.close();
			in.close();
			fin.close();
			fout.close();
		} catch (Exception e) {
			
		}
		
	}

}
